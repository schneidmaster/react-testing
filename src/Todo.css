.todo {
  border: solid 1px #999;
  border-radius: 3px;
  margin-bottom: 20px;
  padding: 15px;
}

.todo-title {
  font-family: Helvetica, Arial, sans-serif;
  font-weight: normal;
  margin: 0px;
  display: flex;
  align-items: center;
}

.todo-title svg:hover {
  cursor: pointer;
}

.todo-title span {
  margin-left: 5px;
}

.todo-title-completed {
  text-decoration: line-through;
}
