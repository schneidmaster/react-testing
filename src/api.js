const API_HOST = "http://localhost:3001";

export const getJSON = (path) => {
  return fetch(`${API_HOST}${path}`).then((response) => response.json());
};

export const postJSON = (path, payload) => {
  return fetch(`${API_HOST}${path}`, {
    method: "POST",
    headers: {
      "content-type": "application/json"
    },
    body: JSON.stringify(payload)
  }).then((response) => response.json());
};

export const patchJSON = (path, payload) => {
  return fetch(`${API_HOST}${path}`, {
    method: "PATCH",
    headers: {
      "content-type": "application/json"
    },
    body: JSON.stringify(payload)
  }).then((response) => response.json());
};
