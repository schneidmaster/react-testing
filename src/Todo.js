import React, { Component } from "react";
import classnames from "classnames";
import { FaRegCheckSquare, FaRegSquare } from "react-icons/fa";
import { patchJSON } from "./api";
import "./Todo.css";

export default class Todo extends Component {
  complete = async () => {
    const todo = await patchJSON(`/todos/${this.props.todo.id}`, {
      status: "complete"
    });
    this.props.updateTodo(todo);
  };

  uncomplete = async () => {
    const todo = await patchJSON(`/todos/${this.props.todo.id}`, {
      status: "incomplete"
    });
    this.props.updateTodo(todo);
  };

  render() {
    const { todo, users } = this.props;

    const assignee = users.find((user) => user.id === todo.userId);

    return (
      <div className="todo">
        <h2 className="todo-title">
          {todo.status === "complete" ? (
            <FaRegCheckSquare
              onClick={this.uncomplete}
              role="checkbox"
              aria-checked="true"
            />
          ) : (
            <FaRegSquare
              onClick={this.complete}
              role="checkbox"
              aria-checked="false"
            />
          )}{" "}
          <span
            className={classnames({
              "todo-title-completed": todo.status === "complete"
            })}
          >
            {todo.title}
          </span>
        </h2>
        <p className="todo-assignee">
          <em>Assigned to:</em> {assignee.name}
        </p>
      </div>
    );
  }
}
