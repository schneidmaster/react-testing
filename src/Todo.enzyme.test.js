import React from "react";
import { shallow } from "enzyme";
import Todo from "./Todo";

// Mock the API so we don't make any real network calls;
jest.mock("./api");
import { patchJSON } from "./api";

describe("Todo", () => {
  beforeEach(() => {
    // Clear out our mock between each test, so that calls
    // from one test don't show up in another test. Don't
    // worry about this just yet -- we'll get to using the
    // mock farther down in this file.
    patchJSON.mockClear();
  });

  // Snapshot tests with Enzyme. Shallow-render the
  // component and snapshot it -- the first time the
  // test runs, it will create a snapshot file in
  // __snapshots__/Todo.enzyme.test.js.snap, and then
  // subsequent runs will compare against that snapshot.
  // When the component markup legitimately changes, you
  // can run the test command with `-u` to update the
  // snapshots.
  //
  // Word of caution: don't let snapshot tests become a
  // replacement for testing functionality. It's easy to
  // write a test suite with 10 snapshots for all the possible
  // component states, but then the snapshots will break
  // every time you make an intentional change and you have
  // to update them all -- and it's easy to casually do this
  // without inspecting all the changes. It's fine to use a
  // few snapshots to ensure rendering doesn't change
  // unexpectedly, but you should have more specific tests
  // asserting the core functionality (more on that below).
  it("renders correctly when incomplete", () => {
    const todo = {
      id: 1,
      title: "Build a racecar 🏎",
      status: "incomplete",
      userId: 1
    };
    const users = [
      {
        id: 1,
        name: "Joe User"
      }
    ];
    const wrapper = shallow(<Todo todo={todo} users={users} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when complete", () => {
    const todo = {
      id: 1,
      title: "Build a racecar 🏎",
      status: "complete",
      userId: 1
    };
    const users = [
      {
        id: 1,
        name: "Joe User"
      }
    ];
    const wrapper = shallow(<Todo todo={todo} users={users} />);
    expect(wrapper).toMatchSnapshot();
  });

  // Using Enzyme finders for a more specific test around which
  // icon renders.
  it("renders empty checkbox when incomplete", () => {
    const todo = {
      id: 1,
      title: "Build a racecar 🏎",
      status: "incomplete",
      userId: 1
    };
    const users = [
      {
        id: 1,
        name: "Joe User"
      }
    ];
    const wrapper = shallow(<Todo todo={todo} users={users} />);
    expect(wrapper.find("FaRegSquare")).toHaveLength(1);
  });

  it("renders checked checkbox when complete", () => {
    const todo = {
      id: 1,
      title: "Build a racecar 🏎",
      status: "complete",
      userId: 1
    };
    const users = [
      {
        id: 1,
        name: "Joe User"
      }
    ];
    const wrapper = shallow(<Todo todo={todo} users={users} />);
    expect(wrapper.find("FaRegCheckSquare")).toHaveLength(1);
  });

  // Let's test out the functionality when we click the checkbox.
  it("can be checked when incomplete", async () => {
    const todo = {
      id: 1,
      title: "Build a racecar 🏎",
      status: "incomplete",
      userId: 1
    };
    const users = [
      {
        id: 1,
        name: "Joe User"
      }
    ];
    const updateTodo = jest.fn();
    const wrapper = shallow(
      <Todo todo={todo} users={users} updateTodo={updateTodo} />
    );

    // Let's make patchJSON return an updated todo with a
    // complete status.
    const updatedTodo = { ...todo, status: "complete" };
    patchJSON.mockImplementation(() => {
      return new Promise((resolve, reject) => {
        resolve(updatedTodo);
      });
    });

    // Note this test is `async`, and `await`s its change --
    // that lets the async method resolve.
    //
    // Enzyme shallow wrappers have a `.find` method which is
    // very flexible -- it can take a component name or a
    // DOM selector. Enzyme also has more finder methods which
    // I'll omit for brevity but are pretty well documented.
    await wrapper.find("FaRegSquare").simulate("click");

    // The component should have made an API call to update
    // the server, and then invoked updateTodo so the parent
    // knows to update its data.
    expect(patchJSON).toBeCalledWith(`/todos/1`, {
      status: "complete"
    });
    expect(updateTodo).toBeCalledWith(updatedTodo);
  });

  // Here's a thing you *can* do -- but never, ever should.
  it("can be checked when method is invoked", async () => {
    const todo = {
      id: 1,
      title: "Build a racecar 🏎",
      status: "incomplete",
      userId: 1
    };
    const users = [
      {
        id: 1,
        name: "Joe User"
      }
    ];
    const updateTodo = jest.fn();
    const wrapper = shallow(
      <Todo todo={todo} users={users} updateTodo={updateTodo} />
    );

    // Let's make patchJSON return an updated todo with a
    // complete status.
    const updatedTodo = { ...todo, status: "complete" };
    patchJSON.mockImplementation(() => {
      return new Promise((resolve, reject) => {
        resolve(updatedTodo);
      });
    });

    // Invoke the method to complete the todo.
    await wrapper.instance().complete();

    // Same as above.
    expect(patchJSON).toBeCalledWith(`/todos/1`, {
      status: "complete"
    });
    expect(updateTodo).toBeCalledWith(updatedTodo);
  });

  // Why shouldn't you do this? Because you are breaking
  // the "4th wall" and testing the implementation details
  // of your component. Tests should validate the component'
  // external behavior, from the perspective of a user or
  // maybe another component.
  //
  // Imagine we refactor this component to use React hooks
  // (aside: guess what's next in the talk!). Externally,
  // the component should behave identically to how it did
  // before -- but this test will break because the component
  // no longer has an instance or a `complete` method. The
  // other tests will not break* because they interact with
  // the interface more like a user would.
  //
  // * actually they will, because Enzyme does not support
  // hooks yet. But the principle -- "don't test implementation
  // details" -- applies no matter what library you use; it
  // will make your tests much less brittle and much more
  // effective at helping guide you through refactors.
});
