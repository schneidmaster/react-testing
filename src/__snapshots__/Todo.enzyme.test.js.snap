// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo renders correctly when complete 1`] = `
<div
  className="todo"
>
  <h2
    className="todo-title"
  >
    <FaRegCheckSquare
      aria-checked="true"
      onClick={[Function]}
      role="checkbox"
    />
     
    <span
      className="todo-title-completed"
    >
      Build a racecar 🏎
    </span>
  </h2>
  <p
    className="todo-assignee"
  >
    <em>
      Assigned to:
    </em>
     
    Joe User
  </p>
</div>
`;

exports[`Todo renders correctly when incomplete 1`] = `
<div
  className="todo"
>
  <h2
    className="todo-title"
  >
    <FaRegSquare
      aria-checked="false"
      onClick={[Function]}
      role="checkbox"
    />
     
    <span
      className=""
    >
      Build a racecar 🏎
    </span>
  </h2>
  <p
    className="todo-assignee"
  >
    <em>
      Assigned to:
    </em>
     
    Joe User
  </p>
</div>
`;
